Issue-4962 Create an s3 archive path with a stable prefix, based on channel

britzl opened the pull request:
`<archive-path>/<channel>/files/*`

britzl committed:
Create an s3 archive path with a stable prefix, based on channel

<archive-path>/<channel>/files/*

britzl committed:
Improved function names and cleaned up to use archive path everywhere

britzl commented on scripts/build.py:
```diff
@@ -1103,16 +1104,14 @@ def build_sdk(self):
 
         platforms = ['x86_64-linux', 'x86_64-darwin', 'win32', 'x86_64-win32', 'armv7-darwin', 'arm64-darwin', 'x86_64-ios', 'armv7-android', 'arm64-android', 'js-web', 'wasm-web']
         for platform in platforms:
-            platform_sdk_url = join(self.archive_path, sha1, 'engine', platform).replace('\\', '/')
```
Unused

britzl commented on scripts/build.py:
```diff
@@ -1649,41 +1651,53 @@ def local_smoke(self):
 
         u = urlparse.urlparse(url)
 
-        if u.netloc == '':
```
upload_file() is only used for S3 uploads. And upload using ssh+scp is not supported anymore.

britzl committed:
Fixed some formatting

britzl committed:
Update build.py

britzl committed:
Added debug prints

britzl committed:
Update build.py

britzl committed:
Trigger ci

britzl committed:
Handle redirects

britzl committed:
Alwyas specify an engine channel for all types of builds

britzl committed:
Added logging when uploading to s3

britzl committed:
trigger ci

britzl committed:
Fixed issue with redirects

britzl committed:
Fix s3 redirects on window

britzl committed:
Another redirect fix

britzl committed:
Adjusted S3 upload url for better lifecycle filters

britzl commented on ci/ci.py:
```diff
@@ -146,7 +146,7 @@ def install(args):
 
 
-def build_engine(platform, with_valgrind = False, with_asan = False, with_vanilla_lua = False, skip_tests = False, skip_codesign = True, skip_docs = False, skip_builtins = False, archive = False):
+def build_engine(platform, with_valgrind = False, with_asan = False, with_vanilla_lua = False, skip_tests = False, skip_codesign = True, skip_docs = False, skip_builtins = False, archive = False, channel = None):
```
Always specify a channel. build.py will default to "stable" if no channel is provided.

britzl commented on ci/ci.py:
```diff
@@ -283,8 +286,15 @@ def release(channel = None):
 
-def build_sdk():
-    call('python scripts/build.py build_sdk')
+def build_sdk(channel = None):
```
Always specify a channel. build.py will default to "stable" if no channel is provided.

britzl commented on editor/tasks/leiningen/util/http_cache.clj:
```diff
@@ -128,6 +128,10 @@
               (doto (File. tmp)
                 (.renameTo dst))
               dst))
+      301 (let [headers (:headers resp)
```
The editor still looks for files in the old location. ie. d.defold.com/archive/sha1/* but these will now only contain redirects/symlinks. We need to handle these and download from the actual location.

britzl commented on scripts/build.py:
```diff
@@ -805,49 +805,49 @@ def archive_engine(self):
             launcher_name = format_exes("launcher", self.target_platform)[0]
             launcherbin = join(bin_dir, launcher_name)
-            self.upload_file(launcherbin, '%s/%s' % (full_archive_path, launcher_name))
+            self.upload_to_archive(launcherbin, '%s/%s' % (full_archive_path, launcher_name))
```
The `upload_to_archive()` function will upload to an S3 path based on the current channel and in a path format that we can easily filter and give different expiration policies:

d.defold.com/beta/files/sha1/*
d.defold.com/stable/files/sha1/*
d.defold.com/alpha/files/sha1/*
...

The upload_to_archive() function will also handle setup of redirects to the old formation (d.defold.com/archive/sha1/*)

britzl committed:
Updated comment

skip-ci

dapetcu21 committed:
Make input stack capacity configurable

JCash reviewed (changes requested):
Looks good, but needs some more tweaks to be configurable with respect to different archives/repos.

JCash commented on ci/ci.py:
```diff
@@ -343,18 +353,18 @@ def main(argv):
         engine_artifacts = args.engine_artifacts or "archived"
     elif branch == "editor-dev":
-        engine_channel = "stable"
+        engine_channel = "editor-alpha"
```
Not sure a bout this name?
What does this mean?
The editor-dev branch doesn't have an engine built?
So how does this map correctly to the stable engine sha1?

JCash commented on ci/ci.py:
```diff
@@ -343,18 +353,18 @@ def main(argv):
         engine_artifacts = args.engine_artifacts or "archived-stable"
     else: # engine dev branch
-        engine_channel = None
+        engine_channel = "engine-dev"
```
If this is "engine-dev", shouldn't the others me engine-beta and engine-master?

britzl replied:
Hmm, I'm thinking that the beta isn't really specific to engine or editor. It's the merged result of engine and editor development.

JCash replied:
Same could be said with "dev" (or the previous name "alpha": http://d.defold.com/alpha), it's the merged result of the editor-dev and dev. The merge is done at the start of every beta.

One could also imagine the editor being released under "editor".
We don't really use any other editor channel.

Here I don't have a strong opinion though.

britzl committed:
Changed folder structure somewhat

britzl commented on ci/ci.py:
```diff
@@ -343,18 +353,18 @@ def main(argv):
         engine_artifacts = args.engine_artifacts
     elif branch and branch.startswith("DEFEDIT-"):
-        engine_channel = None
+        engine_channel = "editor-dev"
```
ci.py has three different channel settings:

1. The `engine_channel` setting is used when calling `archive_engine()`, `archive_sdk()` and `archive_bob()` and it is used when deciding where on S3 to upload (ie d.defold.com/archive/CHANNEL/sha1/*).

2. The `editor_channel` setting is used when building and archiving the editor. 

3. The `release_channel` setting is used by the `release()` function, which is called automatically.

The `DEFEDIT-` branch should never really build the engine, but if it does, we don't want build.py to default to `stable` and archive the engine, bob and sdk as a stable release. That's why I chose `editor-dev` as the path. It could perhaps be renamed `engine-dev` though. What matters is that it is a name different from `stable` and that we are able to set a short expiration date for engine, sdk and bob builds uploaded to d.defold.com/archive/editor-dev/sha1/*.

JCash replied:
Also, since it's called "engine-channel" I associate it with the engine version used/bundled by the editor.

britzl commented on scripts/build.py:
```diff
@@ -58,7 +58,7 @@
 
 CDN_PACKAGES_URL=os.environ.get("DM_PACKAGES_URL", None)
-CDN_UPLOAD_URL="s3://d.defold.com/archive"
+CDN_UPLOAD_URL="s3://d.defold.com"
```
I've updated the PR

britzl committed:
Added missing  --non-constant-id aapt setting to arm64-android

dapetcu21 committed:
Make GetInputStackDefaultCapacity static

JCash committed:
Merge pull request #4971 from dapetcu21/configurable-input-stack-capacity

Make input stack capacity configurable

britzl committed:
Final tweaks to engine_channel based on PR review

skip-ci

britzl committed:
Use forward slash as path separator

The path separator must according to the zip spec be a forward slash. Some implementations and tools accept both, but the bundle-tool.jar validator code looks specifically for forward slash:

https://github.com/google/bundletool/blob/master/src/main/java/com/android/tools/build/bundletool/model/ZipPath.java#L44

britzl committed:
Merge branch 'dev' into Issue-4962-Set-build-expiration-per-channel

britzl committed:
Revert "Use forward slash as path separator"

This reverts commit eb1e8b1195891334985e2393a7c6de25acb05129.

britzl committed:
Retrigger ci
